File: packages/python/plurum/resources/discussions.py
"""
Discussions resource for the Plurum SDK
"""

from __future__ import annotations

from typing import Optional

from plurum._http import HttpClient, AsyncHttpClient


class DiscussionsResource:
    """Synchronous resource for discussion operations"""

    def __init__(self, client: HttpClient):
        self._client = client

    def list_channels(self) -> list[dict]:
        """
        List all discussion channels.

        Returns:
            List of channel objects
        """
        return self._client.get("/api/v1/discussions/channels")

    def list(
        self,
        *,
        channel_slug: Optional[str] = None,
        sort: str = "newest",
        limit: int = 20,
        offset: int = 0,
    ) -> dict:
        """
        List discussion posts with optional filters.

        Args:
            channel_slug: Filter by channel slug
            sort: Sort order ('newest' or 'top')
            limit: Maximum number of results
            offset: Pagination offset

        Returns:
            PostListResponse with items, total, limit, offset, has_more
        """
        params = {"sort": sort, "limit": str(limit), "offset": str(offset)}
        if channel_slug:
            params["channel_slug"] = channel_slug

        query = "&".join(f"{k}={v}" for k, v in params.items())
        return self._client.get(f"/api/v1/discussions/posts?{query}")

    def get(self, short_id: str) -> dict:
        """
        Get a discussion post by short_id with full details.

        Args:
            short_id: The 8-character short_id

        Returns:
            Full post detail with replies
        """
        return self._client.get(f"/api/v1/discussions/posts/{short_id}")

    def create(
        self,
        channel_slug: str,
        title: str,
        body: str,
        *,
        blueprint_identifier: Optional[str] = None,
    ) -> dict:
        """
        Create a new discussion post.

        Args:
            channel_slug: Channel to post in
            title: Post title
            body: Post body content
            blueprint_identifier: Optional linked blueprint (short_id or slug)

        Returns:
            Created post detail
        """
        data: dict = {
            "channel_slug": channel_slug,
            "title": title,
            "body": body,
        }
        if blueprint_identifier:
            data["blueprint_identifier"] = blueprint_identifier

        return self._client.post(
            "/api/v1/discussions/posts", data, requires_auth=True
        )

    def reply(
        self,
        post_short_id: str,
        body: str,
        *,
        parent_reply_id: Optional[str] = None,
    ) -> dict:
        """
        Reply to a discussion post.

        Args:
            post_short_id: Post short_id to reply to
            body: Reply body content
            parent_reply_id: Parent reply ID for nested replies

        Returns:
            Created reply
        """
        data: dict = {"body": body}
        if parent_reply_id:
            data["parent_reply_id"] = parent_reply_id

        return self._client.post(
            f"/api/v1/discussions/posts/{post_short_id}/replies",
            data,
            requires_auth=True,
        )

    def search(
        self,
        query: str,
        *,
        channel_slug: Optional[str] = None,
        limit: int = 10,
    ) -> dict:
        """
        Search discussions using semantic similarity.

        Args:
            query: Natural language search query
            channel_slug: Optional channel filter
            limit: Maximum results

        Returns:
            DiscussionSearchResponse
        """
        params = {"query": query, "limit": str(limit)}
        if channel_slug:
            params["channel_slug"] = channel_slug

        query_str = "&".join(f"{k}={v}" for k, v in params.items())
        return self._client.post(
            f"/api/v1/discussions/search?{query_str}",
        )

    def vote_post(self, post_short_id: str, vote_type: str) -> dict:
        """
        Vote on a post.

        Args:
            post_short_id: Post short_id
            vote_type: 'up' or 'down'

        Returns:
            Vote result with action
        """
        return self._client.post(
            f"/api/v1/discussions/posts/{post_short_id}/vote",
            {"vote_type": vote_type},
            requires_auth=True,
        )

    def vote_reply(self, reply_id: str, vote_type: str) -> dict:
        """
        Vote on a reply.

        Args:
            reply_id: Reply ID
            vote_type: 'up' or 'down'

        Returns:
            Vote result with action
        """
        return self._client.post(
            f"/api/v1/discussions/replies/{reply_id}/vote",
            {"vote_type": vote_type},
            requires_auth=True,
        )


class AsyncDiscussionsResource:
    """Asynchronous resource for discussion operations"""

    def __init__(self, client: AsyncHttpClient):
        self._client = client

    async def list_channels(self) -> list[dict]:
        """List all discussion channels."""
        return await self._client.get("/api/v1/discussions/channels")

    async def list(
        self,
        *,
        channel_slug: Optional[str] = None,
        sort: str = "newest",
        limit: int = 20,
        offset: int = 0,
    ) -> dict:
        """List discussion posts with optional filters."""
        params = {"sort": sort, "limit": str(limit), "offset": str(offset)}
        if channel_slug:
            params["channel_slug"] = channel_slug

        query = "&".join(f"{k}={v}" for k, v in params.items())
        return await self._client.get(f"/api/v1/discussions/posts?{query}")

    async def get(self, short_id: str) -> dict:
        """Get a discussion post by short_id."""
        return await self._client.get(f"/api/v1/discussions/posts/{short_id}")

    async def create(
        self,
        channel_slug: str,
        title: str,
        body: str,
        *,
        blueprint_identifier: Optional[str] = None,
    ) -> dict:
        """Create a new discussion post."""
        data: dict = {
            "channel_slug": channel_slug,
            "title": title,
            "body": body,
        }
        if blueprint_identifier:
            data["blueprint_identifier"] = blueprint_identifier

        return await self._client.post(
            "/api/v1/discussions/posts", data, requires_auth=True
        )

    async def reply(
        self,
        post_short_id: str,
        body: str,
        *,
        parent_reply_id: Optional[str] = None,
    ) -> dict:
        """Reply to a discussion post."""
        data: dict = {"body": body}
        if parent_reply_id:
            data["parent_reply_id"] = parent_reply_id

        return await self._client.post(
            f"/api/v1/discussions/posts/{post_short_id}/replies",
            data,
            requires_auth=True,
        )

    async def search(
        self,
        query: str,
        *,
        channel_slug: Optional[str] = None,
        limit: int = 10,
    ) -> dict:
        """Search discussions using semantic similarity."""
        params = {"query": query, "limit": str(limit)}
        if channel_slug:
            params["channel_slug"] = channel_slug

        query_str = "&".join(f"{k}={v}" for k, v in params.items())
        return await self._client.post(
            f"/api/v1/discussions/search?{query_str}",
        )

    async def vote_post(self, post_short_id: str, vote_type: str) -> dict:
        """Vote on a post."""
        return await self._client.post(
            f"/api/v1/discussions/posts/{post_short_id}/vote",
            {"vote_type": vote_type},
            requires_auth=True,
        )

    async def vote_reply(self, reply_id: str, vote_type: str) -> dict:
        """Vote on a reply."""
        return await self._client.post(
            f"/api/v1/discussions/replies/{reply_id}/vote",
            {"vote_type": vote_type},
            requires_auth=True,
        )

File: packages/typescript/src/resources/discussions.ts
/**
 * Discussions resource for the Plurum SDK
 */

import type { HttpClient } from "../http.js";

export interface DiscussionChannel {
  id: string;
  slug: string;
  name: string;
  description: string | null;
  icon: string | null;
  postCount: number;
  isDefault: boolean;
}

export interface DiscussionPostSummary {
  id: string;
  shortId: string;
  slug: string;
  channelSlug: string;
  channelName: string;
  title: string;
  body: string;
  status: "active" | "closed" | "hidden";
  replyCount: number;
  upvotes: number;
  downvotes: number;
  score: number;
  author: { id: string; name: string; username: string | null };
  blueprint: { shortId: string; slug: string; title: string } | null;
  createdAt: string;
  updatedAt: string;
}

export interface DiscussionReply {
  id: string;
  body: string;
  author: { id: string; name: string; username: string | null };
  upvotes: number;
  downvotes: number;
  score: number;
  isSolution: boolean;
  parentReplyId: string | null;
  depth: number;
  children: DiscussionReply[];
  createdAt: string;
  updatedAt: string;
}

export interface DiscussionPostDetail extends DiscussionPostSummary {
  replies: DiscussionReply[];
}

export interface PostListResponse {
  items: DiscussionPostSummary[];
  total: number;
  limit: number;
  offset: number;
  hasMore: boolean;
}

export interface DiscussionSearchResponse {
  query: string;
  results: Array<{
    post: DiscussionPostSummary;
    similarity: number;
    keywordRank: number;
    combinedScore: number;
    matchReasons: string[];
  }>;
  totalFound: number;
}

// Helper to convert snake_case response to camelCase
function toCamelCase(obj: unknown): unknown {
  if (Array.isArray(obj)) {
    return obj.map(toCamelCase);
  }
  if (obj !== null && typeof obj === "object") {
    const result: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(obj as Record<string, unknown>)) {
      const camelKey = key.replace(/_([a-z])/g, (_, letter) =>
        letter.toUpperCase()
      );
      result[camelKey] = toCamelCase(value);
    }
    return result;
  }
  return obj;
}

export class DiscussionsResource {
  constructor(private http: HttpClient) {}

  /**
   * List all discussion channels.
   */
  async listChannels(): Promise<DiscussionChannel[]> {
    const data = await this.http.get("/api/v1/discussions/channels");
    return toCamelCase(data) as DiscussionChannel[];
  }

  /**
   * List discussion posts with optional filters.
   */
  async list(params?: {
    channelSlug?: string;
    sort?: "newest" | "top";
    limit?: number;
    offset?: number;
  }): Promise<PostListResponse> {
    const queryParams = new URLSearchParams();
    if (params?.channelSlug) queryParams.set("channel_slug", params.channelSlug);
    if (params?.sort) queryParams.set("sort", params.sort);
    if (params?.limit) queryParams.set("limit", String(params.limit));
    if (params?.offset) queryParams.set("offset", String(params.offset));

    const query = queryParams.toString();
    const path = `/api/v1/discussions/posts${query ? `?${query}` : ""}`;
    const data = await this.http.get(path);
    return toCamelCase(data) as PostListResponse;
  }

  /**
   * Get a discussion post by short_id.
   */
  async get(shortId: string): Promise<DiscussionPostDetail> {
    const data = await this.http.get(`/api/v1/discussions/posts/${shortId}`);
    return toCamelCase(data) as DiscussionPostDetail;
  }

  /**
   * Create a new discussion post.
   */
  async create(params: {
    channelSlug: string;
    title: string;
    body: string;
    blueprintIdentifier?: string;
  }): Promise<DiscussionPostDetail> {
    const data = await this.http.post(
      "/api/v1/discussions/posts",
      {
        channel_slug: params.channelSlug,
        title: params.title,
        body: params.body,
        blueprint_identifier: params.blueprintIdentifier,
      },
      true
    );
    return toCamelCase(data) as DiscussionPostDetail;
  }

  /**
   * Reply to a discussion post.
   */
  async reply(
    postShortId: string,
    params: { body: string; parentReplyId?: string }
  ): Promise<DiscussionReply> {
    const data = await this.http.post(
      `/api/v1/discussions/posts/${postShortId}/replies`,
      {
        body: params.body,
        parent_reply_id: params.parentReplyId,
      },
      true
    );
    return toCamelCase(data) as DiscussionReply;
  }

  /**
   * Search discussions.
   */
  async search(params: {
    query: string;
    channelSlug?: string;
    limit?: number;
  }): Promise<DiscussionSearchResponse> {
    const queryParams = new URLSearchParams();
    queryParams.set("query", params.query);
    if (params.channelSlug) queryParams.set("channel_slug", params.channelSlug);
    if (params.limit) queryParams.set("limit", String(params.limit));

    const data = await this.http.post(
      `/api/v1/discussions/search?${queryParams.toString()}`
    );
    return toCamelCase(data) as DiscussionSearchResponse;
  }

  /**
   * Vote on a post.
   */
  async votePost(
    postShortId: string,
    voteType: "up" | "down"
  ): Promise<{ action: string }> {
    return this.http.post(
      `/api/v1/discussions/posts/${postShortId}/vote`,
      { vote_type: voteType },
      true
    );
  }

  /**
   * Vote on a reply.
   */
  async voteReply(
    replyId: string,
    voteType: "up" | "down"
  ): Promise<{ action: string }> {
    return this.http.post(
      `/api/v1/discussions/replies/${replyId}/vote`,
      { vote_type: voteType },
      true
    );
  }
}
